File: library/src/main/java/io/ashdavies/rxtasks/TaskOnSubscribe.java
package io.ashdavies.rxtasks;

import com.google.android.gms.tasks.Task;

abstract class TaskOnSubscribe<Result, Emitter> {

  private final Task<Result> task;
  private final TaskListenerFactory<Result, Emitter> factory;

  TaskOnSubscribe(Task<Result> task, TaskListenerFactory<Result, Emitter> factory) {
    this.task = task;
    this.factory = factory;
  }

  public void subscribe(Emitter emitter) {
    task.addOnSuccessListener(factory.createOnSuccessListener(emitter));
    task.addOnFailureListener(factory.createOnFailureListener(emitter));
  }
}

File: library/src/main/kotlin/io/ashdavies/rx/rxtasks/TaskOnSubscribe.kt
package io.ashdavies.rx.rxtasks

import com.google.android.gms.tasks.Task
import java.util.concurrent.Executor

internal abstract class TaskOnSubscribe<T, in E>(private val task: Task<T>, private val factory: TaskListenerFactory<T, E>) {

  fun subscribe(emitter: E) {
    val scheduler = factory.scheduler
    if (scheduler == null) {
      task.addOnSuccessListener(factory.createOnSuccessListener(emitter))
      task.addOnFailureListener(factory.createOnFailureListener(emitter))
    } else {
      val executor = Executor { scheduler.scheduleDirect(it) }
      task.addOnSuccessListener(executor, factory.createOnSuccessListener(emitter))
      task.addOnFailureListener(executor, factory.createOnFailureListener(emitter))
    }
  }
}
